Improved Encoder API

Barinzaya opened the pull request:
This PR reworks the implementation of the `encoder` module to accomplish 2 goals:

1. Improve the performance of the `encoder` logic. The existing implementation performs a *lot* of small allocations due to creating and passing around around small `Vec` and `String` objects internally. The reworked implementation passes through a reference to the output, and all data is written directly to the output; no internal buffers are allocated, making performance much faster. The performance improvement from this is significant, improving performance by 3-50x in my results from stress-test benchmarks that I created alongside these changes (sample results below). In a practical project in which I'm using `rosc`, these changes have decreased the amount of time spent encoding OSC packets to about 1/8 of what it was before (from ~150 us per packet to ~18 us) with no code changes in that project.

2. Make the output more flexible. This PR adds an `Output` trait which allows the user control over what is done with the output. `Output` is implemented for `Vec<u8>`, allowing for a packet to easily be encoded into a Vec. For `std`, there is also a `WriteOutput` newtype which allows any `std::io::Seek + std::io::Write` type to be used as an `Output`, allowing OSC data to be encoded directly to a `Cursor`, `File`, etc. These changes also include `encode_into` and `encode_string_into` functions, which allow the user to pass in a mutable reference to any object implementing this `Output` trait, allowing the user to control how the encoded data is handled.

The existing public functions have not had their signatures altered, only new public items have been added. Thus, according to the [SemVer Compatibility guide](https://doc.rust-lang.org/cargo/reference/semver.html), these changes should only qualify as a minor change. Additionally, these changes do not incur a MSRV bump (1.52 is the minimum supported version according to `cargo-msrv`, both with and without these changes). Thus, these changes don't require a major version bump.

This also fixes unit tests not building without the `std` feature. All existing unit tests are passing, both with and without `std`. No additional unit tests have been added.

Benchmark results on my machine for comparison:

| Benchmark                                | `master` (Windows) | `encoder-improvements` (Windows) | Relative Speed (Windows) | `master` (Linux)  | `encoder-improvements` (Linux) | Relative Speed (Linux) |
|------------------------------------------|--------------------|----------------------------------|--------------------------|-------------------|--------------------------------|------------------------|
| bench_encode_args_array                  | 90,150 ± 4,137     | 5,929 ± 51                       | 15.2x                    | 39,945 ± 615      | 4,610 ± 52                     | 8.7x                   |
| bench_encode_args_blob                   | 73,101 ± 989       | 8,940 ± 69                       | 8.2x                     | 32,807 ± 201      | 9,931 ± 85                     | 3.3x                   |
| bench_encode_args_bool                   | 37,263 ± 665       | 4,932 ± 107                      | 7.6x                     | 16,431 ± 179      | 4,846 ± 50                     | 3.4x                   |
| bench_encode_args_double                 | 69,613 ± 1,721     | 5,506 ± 39                       | 12.6x                    | 29,751 ± 131      | 5,366 ± 22                     | 5.5x                   |
| bench_encode_args_float                  | 69,645 ± 1,475     | 5,413 ± 22                       | 12.9x                    | 29,924 ± 336      | 4,974 ± 15                     | 6.0x                   |
| bench_encode_args_int                    | 70,911 ± 6,504     | 5,751 ± 43                       | 12.3x                    | 29,742 ± 280      | 4,264 ± 12                     | 7.0x                   |
| bench_encode_args_long                   | 70,653 ± 1,034     | 5,425 ± 46                       | 13.0x                    | 29,634 ± 88       | 5,064 ± 38                     | 5.9x                   |
| bench_encode_args_nil                    | 37,474 ± 918       | 4,983 ± 47                       | 7.5x                     | 16,188 ± 110      | 4,797 ± 18                     | 3.4x                   |
| bench_encode_args_string                 | 141,297 ± 2,025    | 9,083 ± 50                       | 13.9x                    | 46,884 ± 108      | 9,801 ± 34                     | 4.8x                   |
| bench_encode_bundles                     | 294,400 ± 23,868   | 5,844 ± 48                       | 50.4x                    | 99,883 ± 1,409    | 4,639 ± 16                     | 21.5x                  |
| bench_encode_bundles_into_new_vec        | -                  | 6,825 ± 70                       | -                        | -                 | 6,159 ± 281                    | -                      |
| bench_encode_bundles_into_reused_vec     | -                  | 5,282 ± 52                       | -                        | -                 | 6,152 ± 57                     | -                      |
| bench_encode_huge_bundle                 | 2,932,320 ± 91,668 | 562,780 ± 11,300                 | 5.2x                     | 1,270,594 ± 8,991 | 121,758 ± 1,017                | 10.4x                  |
| bench_encode_huge_bundle_into_new_vec    | -                  | 565,861 ± 13,714                 | -                        | -                 | 131,854 ± 862                  | -                      |
| bench_encode_huge_bundle_into_reused_vec | -                  | 137,702 ± 713                    | -                        | -                 | 132,321 ± 832                  | -                      |
| bench_encode_messages                    | 374,490 ± 3,665    | 10,903 ± 241                     | 34.3x                    | 135,246 ± 949     | 9,637 ± 41                     | 14.0x                  |
| bench_encode_messages_into_new_vec       | -                  | 11,988 ± 55                      | -                        | -                 | 11,539 ± 61                    | -                      |
| bench_encode_messages_into_reused_vec    | -                  | 10,386 ± 54                      | -                        | -                 | 11,355 ± 53                    | -                      |
| bench_encode_nested_bundles              | 8,009 ± 136        | 640 ± 25                         | 12.5x                    | 2,842 ± 25        | 240 ± 6                        | 11.8x                  |

I find the discrepancy between the `encode` benchmarks and the `into_new_vec` benchmarks interesting, because they should be doing almost the exact same thing, and yet the `encode` benchmark is noticeably faster in most cases (except for the `huge` benchmark).

anz-ableton commented:
@Barinzaya That's awesome work you did there 🐎 !

I'll try to find some time in the next days, maybe next week, to get a deeper look into the changes and give it a proper review.  Thanks in advance!

Barinzaya commented:
Sure thing! Let me know if you have any comments or suggestions.

klingtnet commented:
@Barinzaya That's awesome work you did there 🐎 !

I'll try to find some time in the next days, maybe next week, to get a deeper look into the changes and give it a proper review. Thanks in advance!

Edit: I deleted the original comment, because I accidentally made that using my work account.

Barinzaya commented:
I simplified the `encode_args` benchmarks to simplify the packets being encoded and focus it more on the arguments (was `bundle(1 message(args))`, now just `message(args))`. For some reason, this seems to have improved the performance on the branch further (expected, since there are fewer steps to the encoding), but for some reason didn't seem to have any effect on the benchmark time on `master`, much to my surprise. This makes the improvement numbers look nicer, but also makes me wonder if the benchmarks I added are missing some case that accounts for that difference.

Just wanted to make an open note of it, since it gives me the feeling of fluffing the numbers, which wasn't the intention at all.

klingtnet commented:
@Barinzaya Can you share the command (if there is some) that you used to create the benchmark table from https://github.com/klingtnet/rosc/pull/44#issue-1617160005 ?

Did you use https://github.com/BurntSushi/cargo-benchcmp ?

klingtnet commented on src/encoder.rs:
```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
+pub fn encode(packet: &OscPacket) -> crate::types::Result<Vec<u8>> {
+    let mut bytes = Vec::new();
+    encode_into(packet, &mut bytes)
+        .expect("Failed to write encoded packet into Vec");
```
What do you think of returning the error from `encode_into` here?  This would be consistent with the previous behavior and how `encode_into` works.

Barinzaya replied:
There's a type incompatibility. `encode` has to return the OSC error type, which isn't compatible with `<Vec<u8> as Output>::Err`, which can either be `core::convert::Infallible` (no `std`) or `std::io::Error` (`std`).

It might be worth making a comment noting it, but neither of these can actually happen in this case. `Infallible` by nature, and the `Write` implementation for `Vec<u8>` will only ever return `Ok`.

klingtnet replied:
Got it, thank you for the explanation.  In this case it's fine to panic if it should ever error out, which might only happen if there's no more memory to allocate I guess.

klingtnet commented on src/encoder.rs:
```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
-fn encode_bundle(bundle: &OscBundle) -> Result<Vec<u8>> {
-    let mut bundle_bytes: Vec<u8> = Vec::new();
-    bundle_bytes.extend(encode_string("#bundle".to_string()).into_iter());
+    let padding = pad(written as u64 + 1) as usize - written;
```
Can you explain why we do the `+1` in `pad(written as u64 + 1)`?  Maybe it's obvious but I did not really get why we need to do that.

klingtnet replied:
Is it because of a string `null` byte?

klingtnet commented on src/encoder.rs:
```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
+                let msg_len = encode_message(m, &mut NullOutput).unwrap();
+                out.reserve(4 + msg_len)?;
+                written += out.write(&(msg_len as u32).to_be_bytes())?;
+                written += encode_message(m, out)?;
```
It's not ideal that we have to encode the message twice here, first to get the length just by discarding the encoding output, and second for the actual encoding.  How much effort do you think would it be to encode just once?

Note that I don't see this as a hard requirement, because this is something that could be done as a follow-up.

klingtnet replied:
I agree, that additionally requiring the `Seek` trait would limit the general usability.

> Generics are monomorphized, so encoding into a NullOutput should effectively be compiled into just calculating and returning the size.

Good point, that should actually be the case.

> If you've got any suggestions on a better way to do this, feel free.

Not yet, but thank you for pointing out that a future change might be a breaking one.  So, I'd like to not merge it now, and instead reflect on this in a couple of days again.   Your performance argument is valid, it might not be worth to optimize this.

Barinzaya replied:
Thinking on this, I see two possible approaches for improving this that could theoretically improve performance, while still allowing the output to be written in a single pass (i.e. without seeking):

1. Encode messages/bundles into a temporary Vec, then write out the length and the message. This would prevent the necessary double-traversal of the packet hierarchy, but would reintroduce a temporary buffer. Using them in this limited capacity could prove beneficial.
2. Memoize the size calculation (if still doing a size pre-pass) or packet encoding (if using buffered encoding a la point 1). This could potentially help with deep hierarchies in both cases, as each OscPacket would only ever be encoded once. This would also require a heap allocation, but for a different purpose (probably using a `HashMap<*const OscPacket, usize>` or something along those lines). I think the biggest improvement from this would be with a deep hierarchy (e.g. `bundle(bundle(bundle(...)))`) due to the O(n^2) visits of the current approach, so I may add an additional benchmark for that as well.

I'll try making these changes and benchmarking them to see whether either of these changes could provide additional performance.

I'll also add here that it occurred to me that unless explicitly replaced (e.g. with `mimalloc` or `jemalloc`), Rust uses the system allocator. Thus, the operating system can have a large impact on the performance of allocation. The numbers currently in the table in the PR description were run on a Windows installation. To compare, I gave the benchmarks a preliminary run on my Linux dual-boot as well (same hardware). The overall performance improvement still looks pretty significant across the board, but it does appear that allocation isn't quite as costly on Linux. The benchmarks that reuse a Vec were nearly the same as on Windows , while all other benchmarks (which allocated new Vecs) were significantly faster than in my original runs on Windows.

I'll be updating the table in the PR description with columns for runtimes on Linux on the same hardware when I get a chance, but I may not be able to get to either task (points above or Linux benchmarks) until tomorrow.

Barinzaya replied:
Added a benchmark for nested bundles. The performance is definitely pathological in the case of a deeply-nested hierarchy.

Encoding bundle contents into a temporary Vec results in comparable performance overall, possibly even slightly beneficial depending on the packet in question. In the benchmarks, it was slower in the case with a wide bundle, faster in the case of the huge bundle, and does completely solve the pathological case. The changes were pushed to the `encoder-improvements+temp` branch on my fork. This seems like a good solution.

Memoizing the sizes also solves the pathological case, but performance overall is worse than just using a temporary Vec. This would require an additional dependency anyway for the `no_std` case, as there is no `HashMap` in `alloc` (as noted before), so `hashbrown` would need to be pulled in. The changes were pushed to the `encoder-improvements+memoize` branch on my fork. This solution doesn't seem practical.

Thinking on it more, doing both memoization and encoding into a temporary Vec isn't beneficial; the O(n^2) nature arises from having a separate size pass, so if a temporary encoding is done, memoization is no longer beneficial.

A comparison of the possible solutions:

| Benchmark | `encoder-improvements` | `encoder-improvements+temp` | `encoder-improvements+memoize` |
|---|---|---|---|
| bench_encode_bundles | 6,331 ± 15 | 9,211 ± 45 | 22,880 ± 702 |
| bench_encode_huge_bundle | 223,253 ± 977 | 156,850 ± 2,248 | 254,628 ± 5,354 |
| bench_encode_messages | 11,307 ± 52 | 11,790 ± 34 | 28,292 ± 126 |
| bench_encode_nested_bundles | 3,787,008 ± 12,171 | 1,624 ± 8 | 1,550 ± 21 |

I think encoding bundle contents into a temporary Vec is the best solution. Temporary buffers are good in moderation!

Barinzaya replied:
Any additional thoughts on this?

klingtnet replied:
Sorry for the delay, I'll try to review this again this weekend.

klingtnet replied:
Thank you for the clear and detailed explanation of the potential performance improvements!  There's not much to add from my side, except that I agree with your statement that encoding into a temporary `Vec` looks like the best choice.  It's also simpler than the memoization approach.

I could make a benchmark run on macOS as well if you like, but I guess that the performance will be similar to the Linux results.

The only thing that is left to do, is to squash the commits.  What do you think of squashing everything that's related to benchmarking into one commit and everything performance related into another one?  It must not be exactly two commits at the end, but I think it would be nice to have the whole performance improvement patches in one commit.

Barinzaya replied:
I agree that the MacOS performance is probably similar to the Linux performance, but it certainly can't hurt to check. You don't necessarily have to format it into a table like I have been (it's a pain to do manually and I still haven't created any kind of shortcut) but a quick check that there aren't any regressions would be good.

As far as commits go, squash away. I definitely tended towards a "commit early and often" approach here, so the number of commits is a bit high. I'm OK with a benchmarks commit and a code changes commit, though if you do the benchmarks first the `encode_into` benchmarks will break. Changes then benchmarks should be good, though.

Thank you for reviewing and thank you for the crate!

klingtnet replied:
> You don't necessarily have to format it into a table like I have been (it's a pain to do manually and I still haven't created any kind of shortcut) but a quick check that there aren't any regressions would be good.

There definitely is no regression, but I also did not expect one.

Benchmark for first commit of this branch  (36b080f): [bench-36b080f.log](https://github.com/klingtnet/rosc/files/11079170/bench-36b080f.log)
Benchmark for the latest commit of this branch (b7b18bc): [bench-b7b18bc.log](https://github.com/klingtnet/rosc/files/11079171/bench-b7b18bc.log)

If you need any other benchmarks, please let me know.  They were done with Rust `1.70.0-nightly` on a M1 MacBook Air.

>  I definitely tended towards a "commit early and often" approach here, so the number of commits is a bit high.

That's totally fine and also the way I usually work.  But, eventually we only care about the end result, without all the intermediate steps.  You're also free to keep any intermediate step/commit where you think it's worth keeping it (for reference, future investigation, debugging, etc).  Any cleanups, refactorings etc. should be squashed or fixed up, though.

> Thank you for reviewing and thank you for the crate!

You're welcome.  I'd like to spend more time on the maintenance of this, and some other crates, but I cannot do that as much as I'd like.  So I'm very happy to receive PRs like yours.  This keeps this project alive.

Barinzaya replied:
Sorry for the delay! This slipped my mind on Monday, and now that I'm starting to look at it again, I'm thinking there's a better solution.

I'm starting to think it would be worth bringing back the write-back mechanism on `Output`. I had done that in my original pass of the `Output` trait, as a means of avoiding a temporary allocation with the single-scan encoding, then made it multi-scan to avoid the need for it. Now that we've effectively gone back to single-scan, though, I'm thinking that adding back a write-back mechanism would be a better alternative than doing internal temporary allocations as we have it now. It increases the requirements of the `Output` trait (effectively requiring it to be seekable), but if the alternative is internally allocating a buffer, I feel that should probably be left in the user's hands. The user can very easily encode into a buffer themselves using `encode` and then write to whatever non-seekable output is needed, but for suitable outputs (e.g. appending to an existing buffer) that should make `encode_into` more efficient without introducing the pathological case that nested bundles had with the multi-scan encoding.

This should also be somewhat more efficient than creating internal buffers, as a single buffer can cover the entire packet, rather than one per message/bundle. Going to play around with this a bit more.

Sorry for changing my mind and putting this back in motion after calling it done!

Barinzaya replied:
Implemented this as described above, it seems to have resulted in *very slight* improvements to most benchmarks on Windows. Benchmarks which encode a lot of bundles (`encode_bundles` and `encode_nested_bundles`) improved substantially.

There's a bit of a regression in the huge bundle benchmarks allocating a new vector compared to the temporary buffer implementation, but a decent gain in the benchmark reusing a vector.

Seems like an overall win to me. Will test on Linux as well.

To allow for the use of `Seek + Write` types as `Output`s, I've added a `WriteOutput<W>` newtype, which can be used as a wrapper to allow any `Seek + Write` type to be used as an `Output`, while still providing a `Vec<u8>` implementation.

| Benchmark                                | Temporary Buffer (Windows) | Write-back (Windows) | Temporary Buffer (Linux) |
|------------------------------------------|----------------------------|----------------------|--------------------------|
| bench_encode_args_array                  | 6,683 ± 138                | 5,929 ± 51           | 5,220 ± 53               |
| bench_encode_args_blob                   | 10,401 ± 676               | 8,940 ± 69           | 10,098 ± 43              |
| bench_encode_args_bool                   | 5,426 ± 36                 | 4,932 ± 107          | 4,992 ± 28               |
| bench_encode_args_double                 | 6,078 ± 77                 | 5,506 ± 39           | 5,104 ± 31               |
| bench_encode_args_float                  | 5,875 ± 68                 | 5,413 ± 22           | 4,760 ± 58               |
| bench_encode_args_int                    | 5,944 ± 39                 | 5,751 ± 43           | 4,847 ± 33               |
| bench_encode_args_long                   | 5,983 ± 176                | 5,425 ± 46           | 5,200 ± 52               |
| bench_encode_args_nil                    | 5,237 ± 35                 | 4,983 ± 47           | 4,749 ± 28               |
| bench_encode_args_string                 | 10,148 ± 95                | 9,083 ± 50           | 9,650 ± 41               |
| bench_encode_bundles                     | 10,044 ± 82                | 5,844 ± 48           | 8,759 ± 104              |
| bench_encode_bundles_into_new_vec        | 11,236 ± 155               | 6,825 ± 70           | 10,076 ± 49              |
| bench_encode_bundles_into_reused_vec     | 9,983 ± 46                 | 5,282 ± 52           | 9,712 ± 144              |
| bench_encode_huge_bundle                 | 511,412 ± 212,302          | 562,780 ± 11,300     | 144,198 ± 1,989          |
| bench_encode_huge_bundle_into_new_vec    | 435,239 ± 17,202           | 565,861 ± 13,714     | 151,398 ± 673            |
| bench_encode_huge_bundle_into_reused_vec | 162,223 ± 2,096            | 137,702 ± 713        | 153,435 ± 3,161          |
| bench_encode_messages                    | 12,695 ± 287               | 10,903 ± 241         | 11,017 ± 49              |
| bench_encode_messages_into_new_vec       | 15,146 ± 310               | 11,988 ± 55          | 12,697 ± 68              |
| bench_encode_messages_into_reused_vec    | 13,728 ± 175               | 10,386 ± 54          | 12,343 ± 86              |
| bench_encode_nested_bundles              | 4,561 ± 82                 | 640 ± 25             | 1,530 ± 23               |

klingtnet commented on src/encoder.rs:
```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
-    let mut bundle_bytes: Vec<u8> = Vec::new();
-    bundle_bytes.extend(encode_string("#bundle".to_string()).into_iter());
+    let padding = pad(written as u64 + 1) as usize - written;
+    written += out.write(&[0u8; 4][..padding])?;
```
I think three bytes should be enough, because if it's 4 then the bytes are already padded.

```suggestion
    written += out.write(&[0u8; 3][..padding])?;
```

klingtnet reviewed (changes requested):
I really like how the new code looks like, especially the idea with the `Output` trait ❤️ 

There're just some minor things to discuss, but nothing is a show stopper.  So, we should be easily able to get this merged.

Thanks again!

Barinzaya commented on benches/encoder_bench.rs:
```diff
@@ -0,0 +1,359 @@
+#![feature(test)]
```
Will do. I overlooked the line endings in creating the file.

Barinzaya commented on benches/encoder_bench.rs:
```diff
@@ -0,0 +1,359 @@
+
+#[bench]
+fn bench_encode_args_array(b: &mut Bencher) {
+	let packet = OscPacket::Message(OscMessage {
```
Sure, I can do that.

Barinzaya commented on benches/encoder_bench.rs:
```diff
@@ -0,0 +1,359 @@
+
+#[bench]
+fn bench_encode_args_blob(b: &mut Bencher) {
+	let packet = OscPacket::Message(OscMessage {
```
Will fix (and run `fmt` in general). Apparently there was a mix of tabs and spaces here for some reason.

Barinzaya commented on benches/encoder_bench.rs:
```diff
@@ -0,0 +1,359 @@
+}
+
+#[bench]
+fn bench_encode_bundles_into_new(b: &mut Bencher) {
```
Will do.

Barinzaya commented on benches/encoder_bench.rs:
```diff
@@ -0,0 +1,359 @@
+}
+
+#[bench]
+fn bench_encode_bundles_into_reuse(b: &mut Bencher) {
```
Will also do.

Barinzaya commented on src/encoder.rs:
````diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
 /// assert!(encoder::encode(&packet).is_ok())
 /// ```
-pub fn encode(packet: &OscPacket) -> Result<Vec<u8>> {
+pub fn encode(packet: &OscPacket) -> crate::types::Result<Vec<u8>> {
````
Correct. `std::result::Result` is used more than `osc::types::Result` now, so I made it the exception. I can import `osc::type::Result as OscResult` or something if you'd prefer.

klingtnet replied:
I think it's a sensible change, just wanted to know if my assumption was correct.

Barinzaya commented:
As far as the benchmark comparison table, I created them manually. Ran `cargo bench` on `master` and copied the output, then switched to `encoder-improvements` and copied its outputs. The relative speeds were also done manually (old time / new time).

I probably could have at least used a spreadsheet or something, but there were only a few benchmarks when I started and I never bothered to set one up. There might be other, more suitable tools out there too.

Barinzaya commented:
Commits have been squashed.

Unsquashed commits are still available in the `encoder-improvements+unsquashed` branch on the fork's repository.

klingtnet commented:
Looks good to me.  I'm going to merge this in the next few days and then prepare a release.
@Barinzaya Good work 👍🏻

Barinzaya commented:
Added one more commit, which adds an encoder unit test using `encode_into` to write to a Vec via a Cursor, using the `WriteOutput` struct. I feel this is somewhat important, as `Output` implementation for `WriteOutput` is not covered by the existing unit tests.

That makes this 5 commits, one per file touched. I can squash them down a bit further (e.g. combine the commit for the unit test change with the commit for the encoder changes) if preferred.
